File: hooks/python/fixonce_hook.py
"""
FixOnce Python Hook - Production Ready
Never debug the same bug twice.

Safety features:
- Anti-loop guard (prevents recursive error capture)
- Rate limiting (max 10 errors per minute)
- Non-blocking sends (uses threading)
- Graceful degradation (silent fail on errors)
"""

import sys
import traceback
import linecache
import json
import threading
import time
from datetime import datetime
from collections import deque
from typing import Optional, Dict, Any

# Configuration
FIXONCE_SERVER = "http://localhost:5000/api/log_error"
ENABLED = True
MAX_VAR_LENGTH = 200
CONTEXT_LINES = 5
MAX_ERRORS_PER_MINUTE = 10  # Rate limiting

# Internal state
_original_excepthook = sys.excepthook
_is_sending = False  # Anti-loop guard
_error_timestamps: deque = deque(maxlen=MAX_ERRORS_PER_MINUTE)
_lock = threading.Lock()


def _rate_limit_check() -> bool:
    """Check if we're within rate limits. Returns True if OK to send."""
    now = time.time()
    with _lock:
        # Remove timestamps older than 60 seconds
        while _error_timestamps and (now - _error_timestamps[0]) > 60:
            _error_timestamps.popleft()

        # Check if we have room
        if len(_error_timestamps) >= MAX_ERRORS_PER_MINUTE:
            return False

        _error_timestamps.append(now)
        return True


def _safe_repr(value, max_length=MAX_VAR_LENGTH) -> str:
    """Safely convert a value to string, truncating if needed."""
    try:
        repr_str = repr(value)
        if len(repr_str) > max_length:
            return repr_str[:max_length] + "... [truncated]"
        return repr_str
    except Exception:
        return "<unrepresentable>"


def _extract_locals(frame) -> Dict[str, str]:
    """Extract local variables from frame, with safety limits."""
    if not frame:
        return {}

    safe_locals = {}
    try:
        for key, value in frame.f_locals.items():
            if key.startswith('__'):
                continue
            if hasattr(value, '__module__') and hasattr(value, '__call__'):
                safe_locals[key] = f"<function {key}>"
                continue
            safe_locals[key] = _safe_repr(value)

        if len(safe_locals) > 20:
            keys = list(safe_locals.keys())[:20]
            safe_locals = {k: safe_locals[k] for k in keys}
            safe_locals["__note__"] = "... more variables truncated"

    except Exception as e:
        safe_locals["__error__"] = f"Could not extract locals: {e}"

    return safe_locals


def _extract_code_snippet(filename: str, lineno: int) -> list:
    """Extract code snippet around the error line."""
    snippet = []

    try:
        start_line = max(1, lineno - CONTEXT_LINES)
        end_line = lineno + CONTEXT_LINES + 1

        for i in range(start_line, end_line):
            line = linecache.getline(filename, i)
            if line:
                marker = ">>>" if i == lineno else "   "
                snippet.append(f"{marker} {i:4d} | {line.rstrip()}")
            elif i == lineno:
                snippet.append(f">>> {i:4d} | <source not available>")

    except Exception as e:
        snippet.append(f"Could not extract code: {e}")

    return snippet


def _send_error_async(payload: Dict[str, Any]):
    """Send error to FixOnce server in background thread."""
    global _is_sending

    def _send():
        global _is_sending
        try:
            import requests
            requests.post(
                FIXONCE_SERVER,
                data=json.dumps(payload, ensure_ascii=False).encode('utf-8'),
                headers={
                    'Content-Type': 'application/json; charset=utf-8',
                    'X-FixOnce-Origin': 'python-hook'  # Anti-loop marker
                },
                timeout=2
            )
        except Exception:
            pass  # Silent fail
        finally:
            _is_sending = False

    _is_sending = True
    thread = threading.Thread(target=_send, daemon=True)
    thread.start()


def fixonce_excepthook(exc_type, exc_value, exc_tb):
    """Rich context exception hook for FixOnce."""
    global _is_sending

    # Always call original first
    _original_excepthook(exc_type, exc_value, exc_tb)

    # Safety checks
    if not ENABLED:
        return

    if _is_sending:  # Anti-loop: don't capture errors while sending
        return

    if not _rate_limit_check():  # Rate limiting
        return

    try:
        error_type = exc_type.__name__
        error_message = str(exc_value)

        filename = ""
        lineno = 0
        func_name = ""
        code_snippet = ["<no traceback available>"]
        local_vars = {}

        if exc_tb:
            tb = exc_tb
            while tb.tb_next:
                tb = tb.tb_next

            frame = tb.tb_frame
            lineno = tb.tb_lineno
            filename = frame.f_code.co_filename
            func_name = frame.f_code.co_name

            code_snippet = _extract_code_snippet(filename, lineno)
            local_vars = _extract_locals(frame)

        stack_trace = "".join(traceback.format_exception(exc_type, exc_value, exc_tb))

        payload = {
            "type": f"python.{error_type}",
            "message": error_message[:500],
            "severity": "critical",
            "file": filename,
            "line": lineno,
            "function": func_name,
            "url": f"python://{filename}",
            "timestamp": datetime.now().isoformat(),
            "snippet": code_snippet,
            "locals": local_vars,
            "stack": stack_trace[:2000],
            "source": "python-hook"
        }

        # Send asynchronously (non-blocking)
        _send_error_async(payload)

    except Exception:
        pass  # Never crash the crash handler


def install():
    """Install the FixOnce error hook."""
    sys.excepthook = fixonce_excepthook
    print("[FixOnce] Python hook installed - Never debug the same bug twice")


def uninstall():
    """Restore the original exception hook."""
    sys.excepthook = _original_excepthook
    print("[FixOnce] Python hook uninstalled")


# Convenience: auto-install when imported
def auto_install():
    """Call this at the start of your main file."""
    install()


# Make import fixonce_hook work as auto-install
if __name__ != "__main__":
    # Auto-install on import (can be disabled by setting ENABLED = False before import)
    pass

File: hooks/node/fixonce-hook.js
/**
 * FixOnce Node.js Hook - Production Ready
 * Never debug the same bug twice.
 *
 * Safety features:
 * - Anti-loop guard (prevents recursive error capture)
 * - Rate limiting (max 10 errors per minute)
 * - Non-blocking sends (async with catch)
 * - Graceful degradation (silent fail)
 *
 * Usage:
 *   require('fixonce-hook'); // At the top of your main file
 *   // or
 *   import 'fixonce-hook';
 */

const FIXONCE_SERVER = 'http://localhost:5000/api/log_error';
const MAX_ERRORS_PER_MINUTE = 10;

// Internal state
let isSending = false;
const errorTimestamps = [];
const originalConsoleError = console.error;

/**
 * Rate limit check - returns true if OK to send
 */
function rateLimitCheck() {
    const now = Date.now();

    // Remove timestamps older than 60 seconds
    while (errorTimestamps.length > 0 && (now - errorTimestamps[0]) > 60000) {
        errorTimestamps.shift();
    }

    // Check limit
    if (errorTimestamps.length >= MAX_ERRORS_PER_MINUTE) {
        return false;
    }

    errorTimestamps.push(now);
    return true;
}

/**
 * Send error to FixOnce server (non-blocking)
 */
async function sendError(payload) {
    if (isSending) return; // Anti-loop guard

    isSending = true;

    try {
        // Use dynamic import for fetch in older Node versions
        const fetchFn = globalThis.fetch || (await import('node-fetch')).default;

        await fetchFn(FIXONCE_SERVER, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-FixOnce-Origin': 'node-hook' // Anti-loop marker
            },
            body: JSON.stringify(payload),
            signal: AbortSignal.timeout(2000) // 2 second timeout
        });
    } catch (e) {
        // Silent fail - never crash the app
    } finally {
        isSending = false;
    }
}

/**
 * Extract stack trace info
 */
function extractStackInfo(error) {
    const stack = error?.stack || '';
    const lines = stack.split('\n');

    let file = '';
    let line = 0;
    let func = '';

    // Find first non-internal stack frame
    for (const stackLine of lines.slice(1)) {
        const match = stackLine.match(/at\s+(?:(.+?)\s+\()?(.+?):(\d+):(\d+)\)?/);
        if (match) {
            func = match[1] || '<anonymous>';
            file = match[2] || '';
            line = parseInt(match[3], 10) || 0;

            // Skip node_modules and internal
            if (!file.includes('node_modules') && !file.startsWith('node:')) {
                break;
            }
        }
    }

    return { file, line, func, stack: stack.substring(0, 2000) };
}

/**
 * Intercept console.error
 */
console.error = function (...args) {
    // Always call original first
    originalConsoleError.apply(console, args);

    // Safety checks
    if (isSending || !rateLimitCheck()) {
        return;
    }

    const message = args
        .map(arg => {
            if (arg instanceof Error) return arg.message;
            if (typeof arg === 'object') {
                try { return JSON.stringify(arg); }
                catch { return String(arg); }
            }
            return String(arg);
        })
        .join(' ')
        .substring(0, 500);

    // Extract error info if first arg is Error
    const firstArg = args[0];
    const isError = firstArg instanceof Error;
    const stackInfo = isError ? extractStackInfo(firstArg) : { file: '', line: 0, func: '', stack: '' };

    const payload = {
        type: isError ? `node.${firstArg.name || 'Error'}` : 'node.console.error',
        message,
        severity: 'error',
        file: stackInfo.file,
        line: stackInfo.line,
        function: stackInfo.func,
        stack: stackInfo.stack,
        timestamp: new Date().toISOString(),
        source: 'node-hook'
    };

    // Send asynchronously (fire and forget)
    sendError(payload);
};

/**
 * Handle uncaught exceptions
 */
process.on('uncaughtException', (error, origin) => {
    // Log to original console first
    originalConsoleError('Uncaught Exception:', error);

    if (isSending || !rateLimitCheck()) {
        // Re-throw to crash as expected
        throw error;
    }

    const stackInfo = extractStackInfo(error);

    const payload = {
        type: `node.${error.name || 'Error'}`,
        message: (error.message || String(error)).substring(0, 500),
        severity: 'critical',
        file: stackInfo.file,
        line: stackInfo.line,
        function: stackInfo.func,
        stack: stackInfo.stack,
        timestamp: new Date().toISOString(),
        source: 'node-hook',
        origin: origin
    };

    sendError(payload);

    // Re-throw to let Node crash normally (don't swallow exceptions)
    setTimeout(() => {
        throw error;
    }, 100);
});

/**
 * Handle unhandled promise rejections
 */
process.on('unhandledRejection', (reason, promise) => {
    originalConsoleError('Unhandled Rejection:', reason);

    if (isSending || !rateLimitCheck()) {
        return;
    }

    const error = reason instanceof Error ? reason : new Error(String(reason));
    const stackInfo = extractStackInfo(error);

    const payload = {
        type: `node.UnhandledRejection`,
        message: (error.message || String(reason)).substring(0, 500),
        severity: 'critical',
        file: stackInfo.file,
        line: stackInfo.line,
        function: stackInfo.func,
        stack: stackInfo.stack,
        timestamp: new Date().toISOString(),
        source: 'node-hook'
    };

    sendError(payload);
});

// Signal that hook is installed
console.log('[FixOnce] Node.js hook installed - Never debug the same bug twice');

// Export for manual control
module.exports = {
    disable: () => {
        console.error = originalConsoleError;
        console.log('[FixOnce] Node.js hook disabled');
    },
    setServer: (url) => {
        FIXONCE_SERVER = url;
    }
};
